model-optimizer/extensions/back/add_outputs_recursive.py

```diff
@@ -0,0 +1,292 @@
+# Copyright (C) 2021 Intel Corporation
+# SPDX-License-Identifier: Apache-2.0
+import numpy as np
+
+from extensions.ops.If import If
+from extensions.ops.loop import Loop
+from extensions.ops.tensor_iterator import TensorIterator
+from mo.back.replacement import BackReplacementPattern
+from mo.front.common.partial_infer.utils import int64_array, is_fully_defined, shape_array
+from mo.front.tf.graph_utils import create_op_node_with_second_input
+from mo.graph.graph import Graph, Node
+from mo.ops.result import Result
+from mo.ops.unsqueeze import Unsqueeze
+
+
+def set_shape_to_port_for_internal_node(cycle_node, internal_id, port_num, iterations_count, axis, need_adjust_port=True):
+    int_node_name = TensorIterator.find_internal_layer_id(cycle_node.body, internal_id)
+    int_node = Node(cycle_node.body, int_node_name)
+    assert int_node.op == 'Result'
+    out_shape = int_node.in_port(0).data.get_shape().copy()
+    # inside cycle node Unsqueeze was added to have the first dimension for concatenating results along it
+    assert len(out_shape) >= 1
+    if axis is not None:
+        out_shape[axis] = iterations_count
+
+    if need_adjust_port:
+        port_num = port_num - len(cycle_node.in_ports())
```

Comment:
as far as I understand this adjustment is not needed, I have tracked functions where this is used and it turns out we can simply use non-adjusted case overall here, i.e. output port index enumerating from 0. It will simplify this logic much.

Reply:
fixed